ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,496 @@
+  * Linting and auto-formatting (Rubocop check/autocorrect but have to customize Rubocop a lot because its defaults are
+    very dogmatic).
+  * Type checking (Steep).
+  * Deployed to RubyGems as separate gems for pure source, x64 linux, x64 darwin, x64 windows (i.e. mingw-ucrt), aarch64
+    linux (i.e. arm64), and arm64 darwin.
+  * CI against supported platforms on oldest and newest Ruby versions supported.
+  * MIT licensed.
+  * Full docs as README.
+  * No need for ruby.temporal.io, https://www.rubydoc.info/gems/temporalio is acceptable.
+  * Full integration test suite.
+
+### Basic Features
+
+By the very first version, the Ruby SDK should have feature parity with every other SDK (except in cases where we have
+decided other features in other SDKs will never be ported). Here's a non-exhaustive list of features we should include
+that are not discussed later:
+
+* Interceptors and an OpenTelemetry implementation
+* Custom metrics and optional metrics buffer for custom processing
+* Typed search attributes, memos, headers, etc
+* API key, RPC metadata, HTTP CONNECT proxy support, etc
+* Loggers with contextual information
+* Full error hierarchy with `Temporalio::Error` as the root
+* ❓ Are there any Ruby libraries so widely adopted they might as well be in the standard library and therefore deserve
+  optional support (e.g. like Go/Python SDK contrib or .NET SDK extensions)?
+
+## Runtime
+
+* `Temporalio::Runtime` is the full Core runtime.
+* Will have class instance variable of `default` that can be set and is lazily created upon first access.
+  * Client (technically connection) accepts a runtime and if unset it defaults to `default`
+* Initialization of the runtime class starts a Ruby thread that just runs the command loop from Rust.
+  * 💭 Why does a runtime need to eat a Ruby thread? The Rust side needs to have a way to make Ruby calls, and only Ruby
+    can start a Ruby thread. So on the Rust side there is essentially a loop with a channel waiting without holding the
+    GVL with things to run on the Ruby side.
+  * 💭 Why do this in a constructor? That is the Ruby way, a constructed object is ready for use.
+  * 💭 Why only a single thread? We expect callbacks to Ruby from Rust to be short (hopefully only a block invocation
+    that is a `Queue.push`) so we don't expect making all of these run serial is an issue. However if it does become an
+    issue, we can use multiple threads and a Rust approach which allows multiple channel consumers.
+* Users can create their own runtime with their own telemetry options and such and pass that to clients.
+  * 💭 Why not a singleton? Technically users may want different runtimes, though hopefully not ever. Still, tests do
+    and it doesn't make sense in many cases to have a singleton that a user creates/initializes.
+
+## Clients
+
+`Temporalio::Client` is a namespace-specific client and `Temporalio::Client::Connection` is a namespace-independent
```

Comment:
> `Temporalio::Client::Connection` is a namespace-independent connection

Are you still holding on to this statement, given the semantic change introduced by Cloud's API Keys? I'm just curious, not arguing either way.

Reply:
Yes. `Connection` is the Rust `Client` and the `CloudOperationsClient` here (as in other SDKs) will leverage the `Connection`. Can see this separation in practice already at `main` of https://github.com/temporalio/sdk-ruby (and in Python SDK and .NET SDK).